File: src/so101_hardware_interface/so101_hardware_interface/hardware_interface_node.py
#!/usr/bin/env python3
"""Bridges the real SO-101 Feetech servo bus to ROS 2.

Wraps LeRobot's SOFollower driver rather than reimplementing a Feetech
bus driver. See package.xml for why this is Python instead of C++.

Contract:
  Subscribes:  /joint_cmd_safe  (so101_msgs/JointCommand)  velocities, rad/s
  Publishes:   /joint_states    (sensor_msgs/JointState)   positions, rad
               /diagnostics     (diagnostic_msgs/DiagnosticArray)

Design notes (see HANDOFF.md Step 2):
  - SOFollower.send_action() takes ABSOLUTE positions in degrees, not
    velocities. This node integrates commanded velocity into a running
    position setpoint per joint (same pattern as manual keyboard-step
    teleop, just with velocity*dt instead of a fixed step).
  - Degree<->radian conversion happens ONLY in this node.
  - Integration uses the REAL measured dt between ticks, not the nominal
    1/control_rate_hz, so a mismatch between configured and actually
    achievable rate cannot silently change commanded speed. Overrun
    ticks are reported via /diagnostics rather than assumed away.
  - On any per-tick exception, the node holds the last commanded
    position (does not integrate this tick, does not crash) per design
    doc §10.
"""

import math
import time

import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import JointState
from diagnostic_msgs.msg import DiagnosticArray, DiagnosticStatus, KeyValue

from so101_msgs.msg import JointCommand

try:
    from lerobot.robots.so_follower.so_follower import (
        SOFollower,
        SOFollowerRobotConfig,
    )
except ImportError as exc:  # pragma: no cover - environment-dependent
    SOFollower = None
    SOFollowerRobotConfig = None
    _IMPORT_ERROR = exc
else:
    _IMPORT_ERROR = None


def _connect_with_retries(robot, tries=3, delay=2.0, logger=None):
    for attempt in range(1, tries + 1):
        try:
            robot.connect(calibrate=False)
            return
        except ConnectionError as e:
            if logger:
                logger.warn(f"Connect attempt {attempt}/{tries} failed: {e}")
            if attempt == tries:
                raise
            time.sleep(delay)


class HardwareInterfaceNode(Node):

    def __init__(self):
        super().__init__('hardware_interface_node')

        if SOFollower is None:
            raise RuntimeError(
                'lerobot SOFollower could not be imported. Is lerobot '
                f'installed in this environment? Original error: {_IMPORT_ERROR}'
            )

        self.declare_parameter('port', '/dev/ttyACM0')
        self.declare_parameter('robot_id', 'vihu2')
        self.declare_parameter('control_rate_hz', 100.0)
        self.declare_parameter('joint_cmd_timeout_ms', 100)
        self.declare_parameter('tick_overrun_warn_ratio', 1.5)
        self.declare_parameter('joint_names', [
            'shoulder_pan', 'shoulder_lift', 'elbow_flex',
            'wrist_flex', 'wrist_roll', 'gripper',
        ])

        self._port = self.get_parameter('port').value
        self._robot_id = self.get_parameter('robot_id').value
        self._control_rate_hz = float(self.get_parameter('control_rate_hz').value)
        self._nominal_dt = 1.0 / self._control_rate_hz
        self._timeout_s = self.get_parameter('joint_cmd_timeout_ms').value / 1000.0
        self._overrun_ratio = self.get_parameter('tick_overrun_warn_ratio').value
        self._joint_names = list(self.get_parameter('joint_names').value)

        # --- connect to the real arm ---
        cfg = SOFollowerRobotConfig(port=self._port, id=self._robot_id)
        self._robot = SOFollower(cfg)
        self.get_logger().info(f'Connecting to SO-101 on {self._port} ...')
        _connect_with_retries(self._robot, logger=self.get_logger())
        self.get_logger().info('Connected.')

        # Seed the running position setpoint from the real arm so the
        # first tick doesn't jump the arm to some default position.
        obs = self._robot.get_observation()
        self._pos_deg = {
            j: obs[f'{j}.pos'] for j in self._joint_names if f'{j}.pos' in obs
        }
        missing = [j for j in self._joint_names if f'{j}.pos' not in obs]
        if missing:
            self.get_logger().error(
                f'Joints not found in SOFollower observation: {missing}. '
                'Check joint_names param against the real robot.'
            )

        self._latest_velocities = {j: 0.0 for j in self._joint_names}
        self._last_cmd_stamp = None  # rclpy Time of last received command
        self._last_tick_time = None  # time.monotonic() of previous tick

        # --- ROS I/O ---
        self._cmd_sub = self.create_subscription(
            JointCommand, '/joint_cmd_safe', self._on_joint_cmd,
            qos_profile_sensor_data,
        )
        self._state_pub = self.create_publisher(
            JointState, '/joint_states', qos_profile_sensor_data,
        )
        self._diag_pub = self.create_publisher(
            DiagnosticArray, '/diagnostics', 10,
        )

        self._timer = self.create_timer(self._nominal_dt, self._tick)

    # ------------------------------------------------------------------
    def _on_joint_cmd(self, msg: JointCommand):
        self._last_cmd_stamp = self.get_clock().now()
        for name, vel in zip(msg.joint_names, msg.velocities):
            if name in self._latest_velocities:
                self._latest_velocities[name] = vel
            else:
                self.get_logger().warn(
                    f'Received velocity for unknown joint "{name}", ignoring.'
                )

    # ------------------------------------------------------------------
    def _publish_diag(self, level, message, values=None):
        arr = DiagnosticArray()
        arr.header.stamp = self.get_clock().now().to_msg()
        status = DiagnosticStatus()
        status.level = level
        status.name = 'hardware_interface_node'
        status.message = message
        if values:
            status.values = [KeyValue(key=k, value=str(v)) for k, v in values.items()]
        arr.status.append(status)
        self._diag_pub.publish(arr)

    # ------------------------------------------------------------------
    def _tick(self):
        now_mono = time.monotonic()
        if self._last_tick_time is None:
            dt = self._nominal_dt
        else:
            dt = now_mono - self._last_tick_time
        self._last_tick_time = now_mono

        if dt > self._overrun_ratio * self._nominal_dt:
            self._publish_diag(
                DiagnosticStatus.WARN,
                'Tick overrun: bus/loop slower than configured control_rate_hz. '
                'Consider lowering control_rate_hz in hardware_params.yaml.',
                {'measured_dt_s': dt, 'nominal_dt_s': self._nominal_dt},
            )

        # Staleness: "user let go" / upstream stopped publishing -> zero velocity.
        stale = (
            self._last_cmd_stamp is None
            or (self.get_clock().now() - self._last_cmd_stamp).nanoseconds
            > self._timeout_s * 1e9
        )
        velocities = (
            {j: 0.0 for j in self._joint_names} if stale else self._latest_velocities
        )

        try:
            for j in self._joint_names:
                if j not in self._pos_deg:
                    continue
                self._pos_deg[j] += math.degrees(velocities.get(j, 0.0)) * dt

            self._robot.send_action(dict(self._pos_deg))
            obs = self._robot.get_observation()

            js = JointState()
            js.header.stamp = self.get_clock().now().to_msg()
            js.name = list(self._joint_names)
            js.position = [
                math.radians(obs.get(f'{j}.pos', float('nan')))
                for j in self._joint_names
            ]
            self._state_pub.publish(js)

        except Exception as e:  # noqa: BLE001 - deliberate broad catch, see §10
            self.get_logger().error(f'Hardware tick failed, holding position: {e}')
            self._publish_diag(
                DiagnosticStatus.ERROR,
                f'Hardware tick exception, holding last commanded position: {e}',
            )
            # Do not re-raise: skip this tick's motion, keep node alive.

    # ------------------------------------------------------------------
    def destroy_node(self):
        try:
            self._robot.disconnect()
        except Exception as e:  # noqa: BLE001
            self.get_logger().warn(f'disconnect() raised, ignoring: {e}')
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = HardwareInterfaceNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/so101_hardware_interface_cpp/src/hardware_interface_node.cpp
// Bridges the real SO-101 Feetech servo bus to ROS 2. C++/ament_cmake
// counterpart of so101_hardware_interface (python) -- see that
// package's docstring for the full rationale on integrate-velocity /
// hold-on-error / measured-dt design. Logic here is intentionally kept
// identical to that version; only the language + Python-embedding layer
// differ.
//
// Contract:
//   Subscribes:  /joint_cmd_safe  (so101_msgs/JointCommand)  velocities, rad/s
//   Publishes:   /joint_states    (sensor_msgs/JointState)   positions, rad
//                /diagnostics     (diagnostic_msgs/DiagnosticArray)

#include <chrono>
#include <cmath>
#include <map>
#include <memory>
#include <optional>
#include <string>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/joint_state.hpp"
#include "diagnostic_msgs/msg/diagnostic_array.hpp"
#include "diagnostic_msgs/msg/diagnostic_status.hpp"
#include "diagnostic_msgs/msg/key_value.hpp"
#include "so101_msgs/msg/joint_command.hpp"

#include "so101_hardware_interface_cpp/so_follower_bridge.hpp"

using namespace std::chrono_literals;

namespace
{
constexpr double kRadToDeg = 180.0 / M_PI;
constexpr double kDegToRad = M_PI / 180.0;
}

class HardwareInterfaceNode : public rclcpp::Node
{
public:
  HardwareInterfaceNode()
  : Node("hardware_interface_node")
  {
    port_ = this->declare_parameter<std::string>("port", "/dev/ttyACM0");
    robot_id_ = this->declare_parameter<std::string>("robot_id", "vihu2");
    venv_site_packages_ = this->declare_parameter<std::string>("venv_site_packages", "");
    control_rate_hz_ = this->declare_parameter<double>("control_rate_hz", 100.0);
    joint_cmd_timeout_ms_ = this->declare_parameter<int>("joint_cmd_timeout_ms", 100);
    overrun_ratio_ = this->declare_parameter<double>("tick_overrun_warn_ratio", 1.5);
    joint_names_ = this->declare_parameter<std::vector<std::string>>(
      "joint_names",
      {"shoulder_pan", "shoulder_lift", "elbow_flex", "wrist_flex", "wrist_roll", "gripper"});

    nominal_dt_ = 1.0 / control_rate_hz_;
    for (const auto & j : joint_names_) {
      latest_velocities_[j] = 0.0;
    }

    bridge_.initInterpreter(venv_site_packages_);
    RCLCPP_INFO(get_logger(), "Connecting to SO-101 on %s ...", port_.c_str());
    bridge_.connect(port_, robot_id_);
    RCLCPP_INFO(get_logger(), "Connected.");

    // Seed running setpoint from the real arm so the first tick doesn't jump it.
    auto obs = bridge_.getObservation();
    for (const auto & j : joint_names_) {
      auto it = obs.find(j + ".pos");
      if (it != obs.end()) {
        pos_deg_[j] = it->second;
      } else {
        RCLCPP_ERROR(
          get_logger(),
          "Joint '%s' not found in SOFollower observation. Check joint_names param.",
          j.c_str());
      }
    }

    rclcpp::QoS sensor_qos = rclcpp::SensorDataQoS();

    cmd_sub_ = this->create_subscription<so101_msgs::msg::JointCommand>(
      "/joint_cmd_safe", sensor_qos,
      std::bind(&HardwareInterfaceNode::onJointCmd, this, std::placeholders::_1));

    state_pub_ = this->create_publisher<sensor_msgs::msg::JointState>(
      "/joint_states", sensor_qos);

    diag_pub_ = this->create_publisher<diagnostic_msgs::msg::DiagnosticArray>(
      "/diagnostics", 10);

    timer_ = this->create_wall_timer(
      std::chrono::duration<double>(nominal_dt_),
      std::bind(&HardwareInterfaceNode::tick, this));
  }

  ~HardwareInterfaceNode() override
  {
    bridge_.disconnect();
  }

private:
  void onJointCmd(const so101_msgs::msg::JointCommand::SharedPtr msg)
  {
    last_cmd_stamp_ = this->get_clock()->now();
    for (size_t i = 0; i < msg->joint_names.size() && i < msg->velocities.size(); ++i) {
      const std::string & name = msg->joint_names[i];
      auto it = latest_velocities_.find(name);
      if (it != latest_velocities_.end()) {
        it->second = msg->velocities[i];
      } else {
        RCLCPP_WARN(
          get_logger(), "Received velocity for unknown joint \"%s\", ignoring.",
          name.c_str());
      }
    }
  }

  void publishDiag(uint8_t level, const std::string & message,
    const std::map<std::string, std::string> & values = {})
  {
    diagnostic_msgs::msg::DiagnosticArray arr;
    arr.header.stamp = this->get_clock()->now();
    diagnostic_msgs::msg::DiagnosticStatus status;
    status.level = level;
    status.name = "hardware_interface_node";
    status.message = message;
    for (const auto & kv : values) {
      diagnostic_msgs::msg::KeyValue keyval;
      keyval.key = kv.first;
      keyval.value = kv.second;
      status.values.push_back(keyval);
    }
    arr.status.push_back(status);
    diag_pub_->publish(arr);
  }

  void tick()
  {
    auto now_steady = std::chrono::steady_clock::now();
    double dt;
    if (!has_last_tick_) {
      dt = nominal_dt_;
      has_last_tick_ = true;
    } else {
      dt = std::chrono::duration<double>(now_steady - last_tick_time_).count();
    }
    last_tick_time_ = now_steady;

    if (dt > overrun_ratio_ * nominal_dt_) {
      publishDiag(
        diagnostic_msgs::msg::DiagnosticStatus::WARN,
        "Tick overrun: bus/loop slower than configured control_rate_hz. "
        "Consider lowering control_rate_hz in hardware_params.yaml.",
        {{"measured_dt_s", std::to_string(dt)}, {"nominal_dt_s", std::to_string(nominal_dt_)}});
    }

    bool stale = true;
    if (last_cmd_stamp_.has_value()) {
      double age_s = (this->get_clock()->now() - last_cmd_stamp_.value()).seconds();
      stale = age_s > (joint_cmd_timeout_ms_ / 1000.0);
    }

    try {
      for (const auto & j : joint_names_) {
        if (pos_deg_.find(j) == pos_deg_.end()) {continue;}
        double vel = stale ? 0.0 : latest_velocities_[j];
        pos_deg_[j] += vel * kRadToDeg * dt;
      }

      bridge_.sendAction(pos_deg_);
      auto obs = bridge_.getObservation();

      sensor_msgs::msg::JointState js;
      js.header.stamp = this->get_clock()->now();
      js.name = joint_names_;
      js.position.reserve(joint_names_.size());
      for (const auto & j : joint_names_) {
        auto it = obs.find(j + ".pos");
        js.position.push_back(
          it != obs.end() ? it->second * kDegToRad : std::nan(""));
      }
      state_pub_->publish(js);

    } catch (const std::exception & e) {
      RCLCPP_ERROR(get_logger(), "Hardware tick failed, holding position: %s", e.what());
      publishDiag(
        diagnostic_msgs::msg::DiagnosticStatus::ERROR,
        std::string("Hardware tick exception, holding last commanded position: ") + e.what());
      // Deliberately not re-thrown: skip this tick's motion, keep node alive (design doc §10).
    }
  }

  // params
  std::string port_;
  std::string robot_id_;
  std::string venv_site_packages_;
  double control_rate_hz_;
  int joint_cmd_timeout_ms_;
  double overrun_ratio_;
  std::vector<std::string> joint_names_;
  double nominal_dt_;

  // state
  so101_hardware_interface_cpp::SoFollowerBridge bridge_;
  std::map<std::string, double> pos_deg_;
  std::map<std::string, double> latest_velocities_;
  std::optional<rclcpp::Time> last_cmd_stamp_;
  std::chrono::steady_clock::time_point last_tick_time_;
  bool has_last_tick_ = false;

  // ROS I/O
  rclcpp::Subscription<so101_msgs::msg::JointCommand>::SharedPtr cmd_sub_;
  rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr state_pub_;
  rclcpp::Publisher<diagnostic_msgs::msg::DiagnosticArray>::SharedPtr diag_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  try {
    auto node = std::make_shared<HardwareInterfaceNode>();
    rclcpp::spin(node);
  } catch (const std::exception & e) {
    RCLCPP_FATAL(rclcpp::get_logger("hardware_interface_node"), "Startup failed: %s", e.what());
    rclcpp::shutdown();
    return 1;
  }
  rclcpp::shutdown();
  return 0;
}
